libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1219 @@
+static ebpf_result_t
+_ebpf_native_initialize_maps(_In_ ebpf_native_map_t* native_maps, _In_ map_entry_t* maps, size_t map_count)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    const int ORIGINAL_FD_OFFSET = 1;
+    for (uint32_t i = 0; i < map_count; i++) {
+        if (maps[i].definition.pinning != PIN_NONE && maps[i].definition.pinning != PIN_GLOBAL_NS) {
+            result = EBPF_INVALID_ARGUMENT;
+            goto Done;
+        }
+        native_maps[i].entry = &maps[i];
+        native_maps[i].original_fd = i + ORIGINAL_FD_OFFSET;
+        native_maps[i].handle = ebpf_handle_invalid;
+        maps[i].address = NULL;
+
+        if (maps[i].definition.pinning == PIN_GLOBAL_NS) {
+            // Construct the pin path.
+            size_t prefix_length = strnlen(DEFAULT_PIN_ROOT_PATH, EBPF_MAX_PIN_PATH_LENGTH);
+            size_t name_length = strnlen_s(maps[i].name, BPF_OBJ_NAME_LEN);
+            if (name_length == 0 || name_length >= BPF_OBJ_NAME_LEN ||
+                prefix_length + name_length + 1 >= EBPF_MAX_PIN_PATH_LENGTH) {
+                result = EBPF_INVALID_ARGUMENT;
+                goto Done;
+            }
+
+            native_maps[i].pin_path.value = ebpf_allocate(prefix_length + name_length + 1);
+            if (native_maps[i].pin_path.value == NULL) {
+                result = EBPF_NO_MEMORY;
+                goto Done;
+            }
+            native_maps[i].pin_path.length = prefix_length + name_length + 1;
+            memcpy(native_maps[i].pin_path.value, DEFAULT_PIN_ROOT_PATH, prefix_length);
+            memcpy(native_maps[i].pin_path.value + prefix_length, "/", 1);
+            memcpy(native_maps[i].pin_path.value + prefix_length + 1, maps[i].name, name_length);
+        } else {
+            native_maps[i].pin_path.value = NULL;
+            native_maps[i].pin_path.length = 0;
+        }
+    }
+
+    // Populate inner map fd.
+    for (uint32_t i = 0; i < map_count; i++) {
+        ebpf_map_definition_in_file_t* definition = &(native_maps[i].entry->definition);
+        int32_t inner_map_original_fd = -1;
+        if (definition->type == BPF_MAP_TYPE_ARRAY_OF_MAPS || definition->type == BPF_MAP_TYPE_HASH_OF_MAPS) {
+            if (definition->inner_map_idx != 0) {
+                inner_map_original_fd = definition->inner_map_idx + ORIGINAL_FD_OFFSET;
+            } else if (definition->inner_id != 0) {
+                for (uint32_t j = 0; j < map_count; j++) {
+                    ebpf_map_definition_in_file_t* inner_definition = &(native_maps[j].entry->definition);
+                    if (inner_definition->id == definition->inner_id && i != j) {
+                        inner_map_original_fd = j + ORIGINAL_FD_OFFSET;
+                        break;
+                    }
+                }
+            }
+        }
+        native_maps[i].inner_map_original_fd = inner_map_original_fd;
+    }
+
+Done:
+    return result;
+}
+
+static inline bool
+_ebpf_native_is_map_in_map(ebpf_native_map_t* map)
```

Comment:
just to verify: is there no existing function to check for is_map_in_map that isn't native specific?

Reply:
Yes, there is none in ebpfcore